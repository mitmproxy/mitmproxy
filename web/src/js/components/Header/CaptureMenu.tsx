import * as React from "react";

CaptureMenu.title = "Capture";

export default function CaptureMenu() {
    //empty component since we don't have an header menu for capture, but we just display the title
    return <></>;
}
