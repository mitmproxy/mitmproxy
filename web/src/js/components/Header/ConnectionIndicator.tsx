import * as React from "react";
import { ConnectionState } from "../../ducks/connection";
import { useAppSelector } from "../../ducks";

export default React.memo(function ConnectionIndicator() {
    const connState = useAppSelector((state) => state.connection.state),
        message = useAppSelector((state) => state.connection.message);

    switch (connState) {
        case ConnectionState.INIT:
            return (
                <span className="connection-indicator init">connecting…</span>
            );
        case ConnectionState.FETCHING:
            return (
                <span className="connection-indicator fetching">
                    fetching data…
                </span>
            );
        case ConnectionState.ESTABLISHED:
            return (
                <span className="connection-indicator established">
                    connected
                </span>
            );
        case ConnectionState.ERROR:
            return (
                <span className="connection-indicator error" title={message}>
                    connection lost
                </span>
            );
        case ConnectionState.OFFLINE:
            return (
                <span className="connection-indicator offline">offline</span>
            );
        default:
            const exhaustiveCheck: never = connState;
            throw "unknown connection state";
    }
});
