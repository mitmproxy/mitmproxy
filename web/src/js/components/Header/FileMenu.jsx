import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import FileChooser from '../common/FileChooser'
import Dropdown, {Divider} from '../common/Dropdown'
import * as flowsActions from '../../ducks/flows'
import * as modalActions from '../../ducks/ui/modal'
import HideInStatic from "../common/HideInStatic";

FileMenu.propTypes = {
    clearFlows: PropTypes.func.isRequired,
    loadFlows: PropTypes.func.isRequired,
    saveFlows: PropTypes.func.isRequired,
}

FileMenu.onNewClick = (e, clearFlows) => {
    e.preventDefault();
    if (confirm('Delete all flows?'))
        clearFlows()
}

export function FileMenu ({clearFlows, loadFlows, saveFlows}) {
     return (
        <Dropdown className="pull-left" btnClass="special" text="mitmproxy">
            <a href="#" onClick={e => FileMenu.onNewClick(e, clearFlows)}>
                <i className="fa fa-fw fa-trash"></i>
                &nbsp;Clear All
            </a>
            <FileChooser
                icon="fa-folder-open"
                text="&nbsp;Open..."
                onOpenFile={file => loadFlows(file)}
            />
            <a href="#" onClick={e =>{ e.preventDefault(); saveFlows();}}>
                <i className="fa fa-fw fa-floppy-o"></i>
                &nbsp;Save...
            </a>

            <HideInStatic>
            <Divider/>
            <a href="http://mitm.it/" target="_blank">
                <i className="fa fa-fw fa-external-link"></i>
                &nbsp;Install Certificates...
            </a>
            </HideInStatic>
        </Dropdown>
    )
}

export default connect(
    null,
    {
        clearFlows: flowsActions.clear,
        loadFlows: flowsActions.upload,
        saveFlows: flowsActions.download,
    }
)(FileMenu)
