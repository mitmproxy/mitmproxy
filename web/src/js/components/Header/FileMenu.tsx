import * as React from "react"
import {useDispatch} from 'react-redux'
import FileChooser from '../common/FileChooser'
import Dropdown, {Divider, MenuItem} from '../common/Dropdown'
import * as flowsActions from '../../ducks/flows'
import HideInStatic from "../common/HideInStatic";
import {useAppSelector} from "../../ducks";


export default React.memo(function FileMenu() {
    const dispatch = useDispatch(), filter = useAppSelector(state => state.flows.filter);
    return (
        <Dropdown className="pull-left special" text="File" options={{"placement": "bottom-start"}}>
            <li>
                <FileChooser
                    icon="fa-folder-open"
                    text="&nbsp;Open..."
                    onClick={
                        // stop event propagation: we must keep the input in DOM for upload to work.
                        e => e.stopPropagation()
                    }
                    onOpenFile={file => {
                        dispatch(flowsActions.upload(file));
                        document.body.click(); // "restart" event propagation
                    }}
                />
            </li>
            <MenuItem onClick={() => location.replace('/flows/dump')}>
                <i className="fa fa-fw fa-floppy-o"/>&nbsp;Save
            </MenuItem>
            <MenuItem onClick={() => location.replace('/flows/dump?filter=' + filter)}>
                <i className="fa fa-fw fa-floppy-o"/>&nbsp;Save filtered
            </MenuItem>
            <MenuItem onClick={() => confirm('Delete all flows?') && dispatch(flowsActions.clear())}>
                <i className="fa fa-fw fa-trash"/>&nbsp;Clear All
            </MenuItem>
            <HideInStatic>
                <Divider/>
                <li>
                    <a href="http://mitm.it/" target="_blank">
                        <i className="fa fa-fw fa-external-link"/>&nbsp;Install Certificates...
                    </a>
                </li>
            </HideInStatic>
        </Dropdown>
    )
});
