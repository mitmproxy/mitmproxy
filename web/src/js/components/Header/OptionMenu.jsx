import React from "react"
import { connect } from "react-redux"
import { EventlogToggle, SettingsToggle } from "./MenuToggle"
import Button from "../common/Button"
import DocsLink from "../common/DocsLink"
import HideInStatic from "../common/HideInStatic";
import * as modalActions from "../../ducks/ui/modal"

OptionMenu.title = 'Options'

function OptionMenu({ openOptions }) {
    return (
        <div>
            <HideInStatic>
                <div className="menu-group">
                    <div className="menu-content">
                        <Button title="Open Options" icon="fa-cogs text-primary"
                                onClick={openOptions}>
                            Edit Options <sup>alpha</sup>
                        </Button>
                    </div>
                    <div className="menu-legend">Options Editor</div>
                </div>

                <div className="menu-group">
                    <div className="menu-content">
                        <SettingsToggle setting="anticache">
                            Strip cache headers <DocsLink resource="overview-features/#anticache"/>
                        </SettingsToggle>
                        <SettingsToggle setting="showhost">
                            Use host header for display
                        </SettingsToggle>
                        <SettingsToggle setting="ssl_insecure">
                            Don't verify server certificates
                        </SettingsToggle>
                    </div>
                    <div className="menu-legend">Quick Options</div>
                </div>
            </HideInStatic>

            <div className="menu-group">
                <div className="menu-content">
                    <EventlogToggle/>
                </div>
                <div className="menu-legend">View Options</div>
            </div>
        </div>
    )
}

export default connect(
    null,
    {
        openOptions: () => modalActions.setActiveModal('OptionModal')
    }
)(OptionMenu)
