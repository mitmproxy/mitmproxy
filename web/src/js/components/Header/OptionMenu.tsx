import * as React from "react";
import { CommandBarToggle, EventlogToggle, OptionsToggle } from "./MenuToggle";
import Button from "../common/Button";
import DocsLink from "../common/DocsLink";
import HideInStatic from "../common/HideInStatic";
import * as modalActions from "../../ducks/ui/modal";
import { useAppDispatch } from "../../ducks";

OptionMenu.title = "Options";

export default function OptionMenu() {
    const dispatch = useAppDispatch();
    const openOptions = () => modalActions.setActiveModal("OptionModal");

    return (
        <div>
            <HideInStatic>
                <div className="menu-group">
                    <div className="menu-content">
                        <Button
                            title="Open Options"
                            icon="fa-cogs text-primary"
                            onClick={() => dispatch(openOptions())}
                        >
                            Edit Options <sup>alpha</sup>
                        </Button>
                    </div>
                    <div className="menu-legend">Options Editor</div>
                </div>

                <div className="menu-group">
                    <div className="menu-content">
                        <OptionsToggle name="anticache">
                            Strip cache headers{" "}
                            <DocsLink resource="overview-features/#anticache" />
                        </OptionsToggle>
                        <OptionsToggle name="showhost">
                            Use host header for display
                        </OptionsToggle>
                        <OptionsToggle name="ssl_insecure">
                            Don't verify server certificates
                        </OptionsToggle>
                    </div>
                    <div className="menu-legend">Quick Options</div>
                </div>
            </HideInStatic>

            <div className="menu-group">
                <div className="menu-content">
                    <EventlogToggle />
                    <CommandBarToggle />
                </div>
                <div className="menu-legend">View Options</div>
            </div>
        </div>
    );
}
