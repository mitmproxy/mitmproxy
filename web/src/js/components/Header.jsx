import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import classnames from 'classnames'
import MainMenu from './Header/MainMenu'
import OptionMenu from './Header/OptionMenu'
import FileMenu from './Header/FileMenu'
import FlowMenu from './Header/FlowMenu'
import {setActiveMenu} from '../ducks/ui/header'
import ConnectionIndicator from "./Header/ConnectionIndicator"
import HideInStatic from './common/HideInStatic'

class Header extends Component {
    static entries = [MainMenu, OptionMenu]

    handleClick(active, e) {
        e.preventDefault()
        this.props.setActiveMenu(active.title)
    }

    render() {
        const { selectedFlowId, activeMenu} = this.props

        let entries = [...Header.entries]
        if(selectedFlowId)
            entries.push(FlowMenu)

        // Make sure to have a fallback in case FlowMenu is selected but we don't have any flows
        // (e.g. because they are all deleted or not yet received)
        const Active = _.find(entries, (e) => e.title == activeMenu) || MainMenu

        return (
            <header>
                <nav className="nav-tabs nav-tabs-lg">
                    <FileMenu/>
                    {entries.map(Entry => (
                        <a key={Entry.title}
                           href="#"
                           className={classnames({ active: Entry === Active})}
                           onClick={e => this.handleClick(Entry, e)}>
                            {Entry.title}
                        </a>
                    ))}
                    <HideInStatic>
                        <ConnectionIndicator/>
                    </HideInStatic>
                </nav>
                <div>
                    <Active/>
                </div>
            </header>
        )
    }
}

export default connect(
    state => ({
        selectedFlowId: state.flows.selected[0],
        activeMenu: state.ui.header.activeMenu,
    }),
    {
        setActiveMenu,
    }
)(Header)
