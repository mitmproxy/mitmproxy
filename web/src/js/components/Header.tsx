import React, { useState } from "react";
import classnames from "classnames";
import StartMenu from "./Header/StartMenu";
import OptionMenu from "./Header/OptionMenu";
import FileMenu from "./Header/FileMenu";
import FlowMenu from "./Header/FlowMenu";
import ConnectionIndicator from "./Header/ConnectionIndicator";
import HideInStatic from "./common/HideInStatic";
import { useAppSelector } from "../ducks";

interface Menu {
    (): JSX.Element;

    title: string;
}

export default function Header() {
    const selectedFlows = useAppSelector((state) =>
            state.flows.selected.filter((id) => id in state.flows.byId)
        ),
        [ActiveMenu, setActiveMenu] = useState<Menu>(() => StartMenu),
        [wasFlowSelected, setWasFlowSelected] = useState(false);

    let entries: Menu[] = [StartMenu, OptionMenu];
    if (selectedFlows.length > 0) {
        if (!wasFlowSelected) {
            setActiveMenu(() => FlowMenu);
            setWasFlowSelected(true);
        }
        entries.push(FlowMenu);
    } else {
        if (wasFlowSelected) {
            setWasFlowSelected(false);
        }
        if (ActiveMenu === FlowMenu) {
            setActiveMenu(() => StartMenu);
        }
    }

    function handleClick(active: Menu, e) {
        e.preventDefault();
        setActiveMenu(() => active);
    }

    return (
        <header>
            <nav className="nav-tabs nav-tabs-lg">
                <FileMenu />
                {entries.map((Entry) => (
                    <a
                        key={Entry.title}
                        href="#"
                        className={classnames({ active: Entry === ActiveMenu })}
                        onClick={(e) => handleClick(Entry, e)}
                    >
                        {Entry.title}
                    </a>
                ))}
                <HideInStatic>
                    <ConnectionIndicator />
                </HideInStatic>
            </nav>
            <div>
                <ActiveMenu />
            </div>
        </header>
    );
}
