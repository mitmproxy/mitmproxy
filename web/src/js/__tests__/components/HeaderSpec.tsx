import * as React from "react";
import { render, screen, fireEvent } from "../test-utils";
import Header from "../../components/Header";
import OptionMenu from "../../components/Header/OptionMenu";

test("Header", async () => {
    const { asFragment } = render(
        <Header ActiveMenu={OptionMenu} setActiveMenu={jest.fn()} />
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Options"));
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText("Edit Options")).toBeTruthy();

    fireEvent.click(screen.getByText("File"));
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText("Open...")).toBeTruthy();

    fireEvent.click(screen.getByText("File"));
    expect(screen.queryByText("Open...")).toBeNull();

    fireEvent.click(screen.getByText("Capture"));
    expect(asFragment()).toMatchSnapshot();
});
