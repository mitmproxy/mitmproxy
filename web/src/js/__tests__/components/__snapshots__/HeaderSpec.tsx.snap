// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="nav-tabs nav-tabs-lg"
    >
      <a
        class="pull-left special"
        href="#"
      >
        File
      </a>
      <a
        class=""
        href="#"
      >
        Start
      </a>
      <a
        class=""
        href="#"
      >
        Options
      </a>
      <a
        class="active"
        href="#"
      >
        Flow
      </a>
      <span
        class="connection-indicator established"
      >
        connected
      </span>
    </nav>
    <div>
      <div
        class="flow-menu"
      >
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <button
              class="btn btn-default"
              disabled=""
              title="[r]eplay flow"
            >
              <i
                class="fa fa-repeat text-primary"
              />
               Replay
            </button>
            <button
              class="btn btn-default"
              title="[D]uplicate flow"
            >
              <i
                class="fa fa-copy text-info"
              />
               Duplicate
            </button>
            <button
              class="btn btn-default"
              disabled=""
              title="revert changes to flow [V]"
            >
              <i
                class="fa fa-history text-warning"
              />
               Revert
            </button>
            <button
              class="btn btn-default"
              title="[d]elete flow"
            >
              <i
                class="fa fa-trash text-danger"
              />
               Delete
            </button>
            <a
              class=""
              href="#"
            >
              <button
                class="btn btn-default"
                title="mark flow"
              >
                <i
                  class="fa fa-paint-brush text-success"
                />
                 Mark▾
              </button>
            </a>
          </div>
          <div
            class="menu-legend"
          >
            Flow Modification
          </div>
        </div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <a
              class=""
              href="#"
            >
              <button
                class="btn btn-default"
              >
                <i
                  class="fa fa-download"
                />
                 Download▾
              </button>
            </a>
            <a
              class=""
              href="#"
            >
              <button
                class="btn btn-default"
                title="Export flow."
              >
                <i
                  class="fa fa-clone"
                />
                 Export▾
              </button>
            </a>
          </div>
          <div
            class="menu-legend"
          >
            Export
          </div>
        </div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <button
              class="btn btn-default"
              disabled=""
              title="[a]ccept intercepted flow"
            >
              <i
                class="fa fa-play text-success"
              />
               Resume
            </button>
            <button
              class="btn btn-default"
              disabled=""
              title="kill intercepted flow [x]"
            >
              <i
                class="fa fa-times text-danger"
              />
               Abort
            </button>
          </div>
          <div
            class="menu-legend"
          >
            Interception
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header 2`] = `
<DocumentFragment>
  <header>
    <nav
      class="nav-tabs nav-tabs-lg"
    >
      <a
        class="pull-left special"
        href="#"
      >
        File
      </a>
      <a
        class=""
        href="#"
      >
        Start
      </a>
      <a
        class="active"
        href="#"
      >
        Options
      </a>
      <a
        class=""
        href="#"
      >
        Flow
      </a>
      <span
        class="connection-indicator established"
      >
        connected
      </span>
    </nav>
    <div>
      <div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <button
              class="btn btn-default"
              title="Open Options"
            >
              <i
                class="fa fa-cogs text-primary"
              />
               Edit Options 
              <sup>
                alpha
              </sup>
            </button>
          </div>
          <div
            class="menu-legend"
          >
            Options Editor
          </div>
        </div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Strip cache headers 
                <a
                  href="https://docs.mitmproxy.org/stable/overview-features/#anticache"
                  target="_blank"
                >
                  <i
                    class="fa fa-question-circle"
                  />
                </a>
              </label>
            </div>
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Use host header for display
              </label>
            </div>
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Don't verify server certificates
              </label>
            </div>
          </div>
          <div
            class="menu-legend"
          >
            Quick Options
          </div>
        </div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <div
              class="menu-entry"
            >
              <label>
                <input
                  checked=""
                  type="checkbox"
                />
                Display Event Log
              </label>
            </div>
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Display Command Bar
              </label>
            </div>
          </div>
          <div
            class="menu-legend"
          >
            View Options
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header 3`] = `
<DocumentFragment>
  <header>
    <nav
      class="nav-tabs nav-tabs-lg"
    >
      <a
        class="pull-left special open"
        href="#"
      >
        File
      </a>
      <ul
        class="dropdown-menu show"
        style="position: absolute; left: 0px; top: 0px;"
      >
        <li>
          <a
            href="#"
          >
            <i
              class="fa fa-fw fa-folder-open"
            />
             Open...
            <input
              class="hidden"
              type="file"
            />
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            <i
              class="fa fa-fw fa-floppy-o"
            />
             Save...
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            <i
              class="fa fa-fw fa-trash"
            />
             Clear All
          </a>
        </li>
        <li
          class="divider"
          role="separator"
        />
        <li>
          <a
            href="http://mitm.it/"
            target="_blank"
          >
            <i
              class="fa fa-fw fa-external-link"
            />
             Install Certificates...
          </a>
        </li>
      </ul>
      <a
        class=""
        href="#"
      >
        Start
      </a>
      <a
        class="active"
        href="#"
      >
        Options
      </a>
      <a
        class=""
        href="#"
      >
        Flow
      </a>
      <span
        class="connection-indicator established"
      >
        connected
      </span>
    </nav>
    <div>
      <div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <button
              class="btn btn-default"
              title="Open Options"
            >
              <i
                class="fa fa-cogs text-primary"
              />
               Edit Options 
              <sup>
                alpha
              </sup>
            </button>
          </div>
          <div
            class="menu-legend"
          >
            Options Editor
          </div>
        </div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Strip cache headers 
                <a
                  href="https://docs.mitmproxy.org/stable/overview-features/#anticache"
                  target="_blank"
                >
                  <i
                    class="fa fa-question-circle"
                  />
                </a>
              </label>
            </div>
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Use host header for display
              </label>
            </div>
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Don't verify server certificates
              </label>
            </div>
          </div>
          <div
            class="menu-legend"
          >
            Quick Options
          </div>
        </div>
        <div
          class="menu-group"
        >
          <div
            class="menu-content"
          >
            <div
              class="menu-entry"
            >
              <label>
                <input
                  checked=""
                  type="checkbox"
                />
                Display Event Log
              </label>
            </div>
            <div
              class="menu-entry"
            >
              <label>
                <input
                  type="checkbox"
                />
                Display Command Bar
              </label>
            </div>
          </div>
          <div
            class="menu-legend"
          >
            View Options
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
